src/analysis/monotone-analyzer.h

```diff
@@ -11,48 +11,123 @@
 
 namespace wasm::analysis {
 
-template<typename Lattice> struct MonotoneCFGAnalyzer;
-
 // A node which contains all the lattice states for a given CFG node.
-template<typename Lattice>
-struct BlockState : public Visitor<BlockState<Lattice>> {
+template<typename Lattice> class BlockState {
   static_assert(is_lattice<Lattice>);
-  // All states are set to the bottom lattice element in this constructor.
-  BlockState(const BasicBlock* underlyingBlock, Lattice& lattice);
-
-  typename Lattice::Element& getFirstState() { return beginningState; }
-  typename Lattice::Element& getLastState() { return endState; }
-
-  // Transfer function implementation. Modifies the state for a particular
-  // expression type.
-  void visitLocalSet(LocalSet* curr);
-  void visitLocalGet(LocalGet* curr);
-
-  // Executes the transfer function on all the expressions of the corresponding
-  // CFG node.
-  void transfer();
-
-  // Prints out all intermediate states in the block.
-  void print(std::ostream& os);
 
-private:
-  // The index of the block is same as the CFG index.
-  Index index;
+  // CFG node corresponding to this state block.
   const BasicBlock* cfgBlock;
-  // State at beginning of CFG node.
-  typename Lattice::Element beginningState;
-  // State at the end of the CFG node.
-  typename Lattice::Element endState;
-  // Holds intermediate state values.
-  typename Lattice::Element currState;
+  // State at which the analysis flow starts for a CFG. For instance, the ending
+  // state for backward analysis, or the beginning state for forward analysis.
+  typename Lattice::Element inputState;
+
   std::vector<BlockState<Lattice>*> predecessors;
   std::vector<BlockState<Lattice>*> successors;
 
-  friend MonotoneCFGAnalyzer<Lattice>;
+public:
+  // All states are set to the bottom lattice element in this constructor.
+  BlockState(const BasicBlock* underlyingBlock, Lattice& lattice);
+
+  // Accessors.
+  typename Lattice::Element& getInputState() { return inputState; }
+  const BasicBlock* getCFGBlock() const { return cfgBlock; }
```

Comment:
These are only used internally by the monotone analyzer code itself, right? If so, I don't see any benefit to using accessor functions rather than just accessing these members directly.

Reply:
We discussed offline that the `BlockState` contents should be made public, but not be exposed to any class other than the `MonotoneCFGAnalyzer`. One reason is that friending `MonotoneCFGAnalyzer` is cumbersome because it has a `TransferFunction` template that `BlockState` should not be aware of.

Since we are doing this, it is also not necessary to store the predecessors and successors in `BlockState` since we can just get the corresponding index from the underling CFG `BasicBlock` and look up the index in `MonotoneCFGAnalyzer`. I have therefore taken out the predecessors and successors from `BlockState` also.